File: 0152-maximum-product-subarray/0152-maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        if(nums.size()==1) return nums[0];
        priority_queue<int> pq;
        int prod=1,res=nums[0],zc=0;;
        for(auto& i:nums){
            prod*=i;
            if(prod<0){
                int temp = prod;
                if(pq.size()) temp/=pq.top();
                res=max({res,temp,prod});
                pq.push(prod);
            }
            else if(!prod){
                res=max(res,prod);
                prod=1;
                pq=priority_queue<int>();
            }
            else res=max(res,prod);
        }

        return res;

    }
};

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int: 
        res = float("-inf")
        prod = 1 
        min_prod = float("-inf") # for negative prod 
        for val in nums: 
            prod*=val
            #>0
            if prod>0: 
                res = max(res,prod) 
            #<0 
            elif prod<0: 
                min_prod = max(min_prod,prod)
                if prod!=min_prod: 
                    rem = prod//min_prod
                    print(rem)
                    res = max(res,rem) 
                res=max(res,prod)
            #=0
            else: 
                prod = 1
                min_prod = float("-inf")
            res=max(res,val)
        return res
